File: solutions/875.koko-eating-bananas.py
#
# @lc app=leetcode id=875 lang=python3
#
# [875] Koko Eating Bananas
#

# @lc code=start
import math
from typing import List


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        low = 1
        high = max(piles)
        while low < high:
            middle = (low + high) // 2
            spent_hours = sum(math.ceil(float(p) / middle) for p in piles)
            if spent_hours <= h:
                high = middle
            else:
                low = middle + 1

        return high

# @lc code=end

File: solutions/875.koko-eating-bananas.java
/*
 * @lc app=leetcode id=875 lang=java
 *
 * [875] Koko Eating Bananas
 */

// @lc code=start

import java.util.Arrays;

class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int min = 1;
        int max = Arrays.stream(piles).max().orElseThrow();
        while (min < max) {
            int middle = (min + max) >>> 1;
            int spentHours = 0;
            for(int i = 0; i < piles.length; i++) {
                spentHours += Math.ceilDiv(piles[i], middle);
            }
            if (spentHours <= h) {
                max = middle;
            } else {
                min = middle + 1;
            }
        }
        return max;
    }
}
// @lc code=end

